x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java

```diff
@@ -25,14 +25,42 @@
  */
 class FieldValueFetcher {
 
-    private final MappedFieldType fieldType;
-    private final IndexFieldData<?> fieldData;
+    protected final MappedFieldType fieldType;
+    protected final IndexFieldData<?> fieldData;
     private final AbstractRollupFieldProducer rollupFieldProducer;
 
+    private static class RollupFieldProducersFactory {
+
+        static AbstractRollupFieldProducer createProducer(final MappedFieldType fieldType, final String name) {
+            if (fieldType.getMetricType() == null) {
+                return createLabelFieldProducer(fieldType, name);
+            }
+            return createMetricFieldProducer(fieldType, name);
+        }
+
+        private static AbstractRollupFieldProducer createMetricFieldProducer(final MappedFieldType fieldType, final String name) {
+            return switch (fieldType.getMetricType()) {
+                case GAUGE -> new MetricFieldProducer.GaugeMetricFieldProducer(fieldType, name);
+                case COUNTER -> new MetricFieldProducer.CounterMetricFieldProducer(fieldType, name);
+            };
+        }
+
+        private static AbstractRollupFieldProducer createLabelFieldProducer(final MappedFieldType fieldType, final String name) {
+            if (isHistogramField(fieldType)) {
+                return new LabelFieldProducer.HistogramLabelFieldProducer(fieldType, name);
+            }
+            return new LabelFieldProducer.LabelLastValueFieldProducer(fieldType, name);
+        }
+
+        private static boolean isHistogramField(final MappedFieldType fieldType) {
+            return HistogramFieldMapper.CONTENT_TYPE.equals(fieldType.typeName());
```

Comment:
This is why we need the dependency in the Gradle file.

Reply:
Maybe replace with:

```
return "histogram".equals(fieldType.typeName());
```

I think we should avoid non test dependencies if we can.

Reply:
Yes I was looking at a different way of doing it to avoid relying on string comparison...